src/utils/internalLinks.js

```diff
@@ -0,0 +1,117 @@
+/* @flow strict-local */
+import type { Narrow, User, Stream } from '../types';
+import { topicNarrow, streamNarrow, groupNarrow, specialNarrow } from './narrow';
+import { isUrlOnRealm } from './url';
+import { getUserById } from '../users/userHelpers';
+import { transformToEncodedURI } from './string';
+import { NULL_USER } from '../nullObjects';
+
+export const getPathsFromUrl = (url: string = '', realm: string) => {
+  const paths = url
+    .split(realm)
+    .pop()
+    .split('#narrow/')
+    .pop()
+    .split('/');
+
+  if (paths.length > 0 && paths[paths.length - 1] === '') {
+    // url ends with /
+    paths.splice(-1, 1);
+  }
+  return paths;
+};
+
+export const isInternalLink = (url: string, realm: string): boolean =>
+  isUrlOnRealm(url, realm) ? /^(\/#narrow|#narrow)/i.test(url.split(realm).pop()) : false;
+
+export const isMessageLink = (url: string, realm: string): boolean =>
+  isInternalLink(url, realm) && url.includes('near');
+
+type LinkType = 'external' | 'home' | 'group' | 'topic' | 'stream' | 'special';
+
+export const getLinkType = (url: string, realm: string): LinkType => {
+  if (!isInternalLink(url, realm)) {
+    return 'external';
+  }
+
+  const paths = getPathsFromUrl(url, realm);
+
+  if (
+    (paths.length === 2 && paths[0] === 'pm-with')
+    || (paths.length === 4 && paths[0] === 'pm-with' && paths[2] === 'near')
+  ) {
+    return 'group';
+  }
+
+  if (
+    (paths.length === 4 || paths.length === 6)
+    && paths[0] === 'stream'
+    && (paths[2] === 'subject' || paths[2] === 'topic')
+  ) {
+    return 'topic';
+  }
+
+  if (paths.length === 2 && paths[0] === 'stream') {
+    return 'stream';
+  }
+
+  if (paths.length === 2 && paths[0] === 'is' && /^(private|starred|mentioned)/i.test(paths[1])) {
+    return 'special';
+  }
+
+  return 'home';
+};
+
+export const getNarrowFromLink = (
+  url: string,
+  realm: string,
+  users: User[],
+  streams: Stream[],
+): ?Narrow => {
```

Comment:
This is never `undefined`, right? Only `null`, or a `Narrow`.

If that's right, let's write `Narrow | null` to describe the situation precisely. Flow's shorthand `?Narrow` unfortunately means `Narrow | null | void`, which is virtually never what we want -- the reality is almost always something more specific, like `Narrow | null` or `Narrow | void`.

Where this becomes really important is when the main type is `string` or `number` or `boolean`, something that can itself be falsy -- because then `if (foo)` is bad, and if we write `foo: ?Foo` we have to write redundant checks like `if (foo !== null && foo !== undefined)`, when in reality only one of `null` and `undefined` can ever happen. But I think it's simpler to have a blanket policy of writing the specific `Foo | null` or `Foo | void` and never `?Foo`, even when we could get away with `if (foo)`.